### ISSUE TITLE
FLTK crashes on macOS 12 (11?)

### ISSUE THREAD

**Issue by CendioOssman** [fltk/fltk#373]
We're seeing constant crashes when running FLTK 1.3.7 on a M1 processor. FLTK 1.3.5 works just fine though. This is with an x86_64 process being run through Rosetta. It's unknown if a native compiled binary has the same issues.

We're currently debugging this so I'll ask for your patience in more details. I wanted to file the issue right away in case you are considering doing a 1.3.9 and it might be worth waiting for a fix here.

What we can see is that the crash happens in graphics operations, but the stack is pretty devoid of FLTK calls, so there is some delay to things. We're also getting lots of warnings about invalid graphics context when running things under lldb.

46235ff92216339e1a3ff2796df1a83a2e6418c2 touches a lot of the graphics stuff, so that's probably where we'll focus our investigation at first.

**Comment by wyatt8740** [comment]
They're doing a 1.4.0, as far as I can tell.

So I assume this is Mac OS? Does it crash immediately on startup?

Also, [1.3.8 claims M1 support](https://www.fltk.org/articles.php?L825); why don't you try 1.3.8?

**Comment by MatthiasWM** [comment]
We are in the process of releasing 1.4.0 . I am currently extending Fluid and modernising Fl_Preferences. My main machine is a MacBook Pro with M1 and I run native M1 FLTK apps daily without crashes, occasionally Intel FLTK code for verification, also without crashes. Please try 1.4 for your project. It is quite stable already. I also recommend that you generate Universal Binaries when releasing software, so it runs equally well on M1 and Intel.

If the crash persists, it would be easiest for us if you can reduce the issue to a small piece of source code that we can compile locally to reproduce the bug. Maybe getting a backtrace of one of your crashes may already help.

**Comment by CendioOssman** [comment]
Alright, so we've gotten a bit further. We've seen the crash once on x86 now as well. But just once. So it seems to happen more often on M1, but is not exclusive to that platform.

A crash dump can be found here for those interested:

https://community.thinlinc.com/t/thinlinc-4-14-0-beta/288/2

It's not too helpful though as it seems it crashes way after the issue has been triggered.

The path we investigated was instead these warnings that we got whenever a FLTK binary was run under lldb:

```
2022-01-19 12:24:11.563263+0100 unittests[2853:668914] [unittests] CGContextRestoreGState: invalid context 0x600003711080. If you want to see the backtrace, please set CG_CONTEXT_SHOW_BACKTRACE environmental variable.
2022-01-19 12:24:11.563284+0100 unittests[2853:668914] [unittests] CGContextRestoreGState: invalid context 0x600003711080. If you want to see the backtrace, please set CG_CONTEXT_SHOW_BACKTRACE environmental variable.
2022-01-19 12:24:11.563293+0100 unittests[2853:668914] [unittests] CGContextFlush: invalid context 0x600003711080. If you want to see the backtrace, please set CG_CONTEXT_SHOW_BACKTRACE environmental variable.
```

Googling it unfortunately gave mostly hits that this was false errors on some version of iOS. But one place mentioned that this could be because drawing is done outside of `drawRect:`.

A second clue was enabling `CG_CONTEXT_SHOW_BACKTRACE`. It unfortunately didn't give us a full trace, but it did tell us the calling function which was `Fl_X::q_release_context()`.

So we did some more tracing and discovered that the graphics context was grabbed from the system in `drawRect:`, but it was released much later, outside of `drawRect:`. It seems those cleanup calls are enough to trigger the issues.

Next step was seeing if things would work better if we made sure the context is released before `drawRect:` returns. And apparently it already. Provided you have a recent enough SDK. Which we don't have, since we're stuck on an ancient one (for various practical reasons).

Moving that line so it is always included seems to fix the bug. We're still testing as this thing is a bit random. It does get rid of the complaints you get via lldb though.

```diff
diff --git a/src/Fl_cocoa.mm b/src/Fl_cocoa.mm
index 3ce77f95b..c3b7b8abe 100644
--- a/src/Fl_cocoa.mm
+++ b/src/Fl_cocoa.mm
@@ -2474,8 +2475,8 @@ static FLTextInputContext* fltextinputcontext_instance = nil;
       CGImageRelease(img);
     }
   }
-  Fl_X::q_release_context();
 #endif
+  Fl_X::q_release_context();
   if (!through_Fl_X_flush) window->clear_damage();
   through_drawRect = NO;
   fl_unlock_function();
```

This was tested with the current state of the 1.3 branch.  From what I can tell 1.4 still has the same code though.

**Comment by samhed** [comment]
We have finished testing the above fix at our end now, we're fairly certain that it fixes the crash.

**Comment by Albrecht-S** [comment]
Thank you very much for the analysis, thorough testing, and the patch.

@ManoloFLTK What do you think about this patch? It looks very "simple" but I can't tell if it is correct.

**Comment by ManoloFLTK** [comment]
@CendioOssman Many thanks for finding this error and its cure. Can you use a patched FLTK 1.3.8 for macOS? FLTK 1.4 should be out soon and has been fixed too.

**Comment by CendioOssman** [comment]
We've been running a patched FLTK since we found the fix, so we're good. :+1:
I believe @bphinz has also patched his FLTK that is used for the official TigerVNC builds.
